File: DataStructures-Algorithms/8. Trees/Sorted_array_to_BST.c
/*
Convert a Sorted Array into BST

Time Complexity : O(n)
Space Complexity : O(n)
*/

#include<stdio.h>
#include<stdlib.h>

struct node
{
	int data;
	struct node *left,*right;
};

struct node *newNode(int val)
{
	struct node *nn = (struct node*)malloc(sizeof(struct node));
	
	nn->data = val;
	
	nn->left = nn->right = NULL;

	return nn;
}

struct node *makeBST( int *arr, int start, int end) // O(n)
{
	if(start > end)
		return NULL;

	int mid = (start+end)/2;

	struct node *root = newNode(arr[mid]);

	root->left = makeBST(arr,start,mid-1);
	
	root->right = makeBST(arr,mid+1,end);

	return root;
}

void PreOrder(struct node *root)
{
	if(!root)
		return;
	
	printf(" %d",root->data);
	
	PreOrder(root->left);
	
	PreOrder(root->right);
}

int main()
{
	int arr[] = {1, 2, 3, 4, 5, 6, 7};
	
	int n = sizeof(arr)/sizeof(*arr);

	PreOrder(makeBST(arr, 0, n-1));
}

File: interview/src/tree/SortedArrayToBST.java
/**
 * https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
 *
 * Time Complexity : O(N)
 * Space Complexity : O(log(N))
 * */
package tree;

public class SortedArrayToBST extends TreeNode {
        
        private TreeNode sortedArrayToBSTUtil(int[] nums, int start, int end) {
                // base case
                if (start > end) {
                        return null;
                }
                // get mid element and make as root
                int mid = (start + end) / 2;
                TreeNode root = new TreeNode(nums[mid]);
                // recur for left and right subtrees
                root.left = sortedArrayToBSTUtil(nums, start, mid - 1);
                root.right = sortedArrayToBSTUtil(nums, mid + 1, end);
                return root;
        }
        
        private TreeNode sortedArrayToBST(int[] nums) {
                // base case
                if (nums.length == 0) {
                        return null;
                }
                return sortedArrayToBSTUtil(nums, 0, nums.length - 1);
        }
        
        public static void main(String[] args) {
                int nums[] = {-10, -3, 0, 5, 9};
                SortedArrayToBST obj = new SortedArrayToBST();
                obj.preOrder(obj.sortedArrayToBST(nums));
        }
}
